extensions/gson/src/test/groovy/io/jsonwebtoken/gson/io/GsonSerializerTest.groovy

```diff
@@ -0,0 +1,99 @@
+package io.jsonwebtoken.gson.io
+
+import io.jsonwebtoken.lang.Strings
+import org.junit.Test
+
+import static org.easymock.EasyMock.*
+import static org.junit.Assert.*
+import com.google.gson.Gson
+
+class GsonSerializerTest {
+
+    @Test
+    void testDefaultConstructor() {
+        def serializer = new GsonSerializer()
+        assertNotNull serializer.gson
+    }
+
+    @Test
+    void testObjectMapperConstructor() {
+        def customGSON = new Gson()
+        def serializer = new GsonSerializer<>(customGSON)
+        assertSame customGSON, serializer.gson
+    }
+
+    @Test(expected = IllegalArgumentException)
+    void testObjectMapperConstructorWithNullArgument() {
+        new GsonSerializer<>(null)
+    }
+
+    @Test
+    void testByte() {
+        byte[] expected = "120".getBytes(Strings.UTF_8) //ascii("x") = 120
+        byte[] bytes = "x".getBytes(Strings.UTF_8)
+        byte[] result = new GsonSerializer().serialize(bytes[0]) //single byte
+        assertTrue Arrays.equals(expected, result)
+    }
+
+    @Test
+    void testByteArray() { //expect Base64 string by default:
+        byte[] bytes = "hi".getBytes(Strings.UTF_8)
+        String expected = '"aGk="' as String //base64(hi) --> aGk=
+        byte[] result = new GsonSerializer().serialize(bytes)
+        assertEquals expected, new String(result, Strings.UTF_8)
+    }
+
+    @Test
+    void testEmptyByteArray() { //expect Base64 string by default:
+        byte[] bytes = new byte[0]
+        byte[] result = new GsonSerializer().serialize(bytes)
+        assertEquals '""', new String(result, Strings.UTF_8)
+    }
+
+    @Test
+    void testChar() { //expect Base64 string by default:
+        byte[] result = new GsonSerializer().serialize('h' as char)
+        assertEquals "\"h\"", new String(result, Strings.UTF_8)
+    }
+
+    @Test
+    void testCharArray() { //expect Base64 string by default:
+        byte[] result = new GsonSerializer().serialize("hi".toCharArray())
+        assertEquals "\"hi\"", new String(result, Strings.UTF_8)
+    }
+
+    @Test
+    void testSerialize() {
+        byte[] expected = '{"hello":"世界"}'.getBytes(Strings.UTF_8)
+        byte[] result = new GsonSerializer().serialize([hello: '世界'])
+        assertTrue Arrays.equals(expected, result)
+    }
+
+//THIS IS COMMENTED OUT BECAUSE GSN NEVER THROWS ANY SERIALIZATION EXCEPTION    
```

Comment:
You have to re-enable this test @patton73 Code coverage requires it.

Reply:
Well gson does not lauch any exception do I write a fake test about that?

Reply:
The test actually handles that.  It throws an exception by overriding the serialization on line 82

Reply:
Just choose a different exception.  ANy would do.  the OrgJson extension uses IllegalArgumentException:  https://github.com/jwtk/jjwt/blob/73f7f9915baccaa3f3143e549647f818606d49ce/extensions/orgjson/src/test/groovy/io/jsonwebtoken/io/OrgJsonSerializerTest.groovy#L35

Reply:
If i re-enable this the test will never pass since gson will never throw a serialization exception.

Reply:
The following code creates a Mock GSONSerializer that DOES throw an exception.  Now I agree its a bit silly because GSON won't even throw an exception outside of this unit test, but @lhazlewood wanted the try catch there for consistency and to maintain 100% code coverage you need to hit that try catch
```
    @Test
    void testSerializeFailsWithWrappedException() {

        def ex = createMock(Exception)

        expect(ex.getMessage()).andReturn('foo')

        def serializer = new GsonSerializer() {
            @Override
            protected byte[] writeValueAsBytes(Object o) {
                throw ex
            }
        }

        replay ex

        try {
            serializer.serialize([hello: 'world'])
            fail()
        } catch (SerializationException se) {
            assertEquals 'Unable to serialize object: foo', se.getMessage()
            assertSame ex, se.getCause()
        }

        verify ex
    }```

Reply:
Just for curiosity, is there a reason behind the use of groovy to test java instead of junit?

Reply:
@patton73 we are using JUnit as the testing library.  We just use Groovy to write the tests - it's faster to write, allows us to ignore Java visibility modifiers to assert values and state, reflection is far easier, mocking things can be simpler, and it's fun to use.

In other words, it's far simpler and more productive to write tests in Groovy, but we still use JUnit to execute the tests and assertions.  HTH!